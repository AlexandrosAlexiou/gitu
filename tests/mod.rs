use crate::helpers::{clone_and_commit, commit, key, key_code, run, TestContext};
use crossterm::event::KeyCode;
use std::fs;

mod helpers;

#[test]
fn no_repo() {
    let ctx = TestContext::setup_init(60, 20);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn help_menu() {
    let mut ctx = TestContext::setup_init(60, 20);
    ctx.update(&[key('h')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn fresh_init() {
    let mut ctx = TestContext::setup_init(60, 20);
    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn new_file() {
    let mut ctx = TestContext::setup_init(60, 20);
    run(ctx.dir.path(), &["touch", "new-file"]);
    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn unstaged_changes() {
    let mut ctx = TestContext::setup_init(60, 20);
    commit(ctx.dir.path(), "testfile", "testing\ntesttest");
    fs::write(ctx.dir.child("testfile"), "test\ntesttest").expect("error writing to file");

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn staged_file() {
    let mut ctx = TestContext::setup_init(60, 20);
    run(ctx.dir.path(), &["touch", "new-file"]);
    run(ctx.dir.path(), &["git", "add", "new-file"]);
    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn log() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "firstfile", "testing\ntesttest\n");
    commit(ctx.dir.path(), "secondfile", "testing\ntesttest\n");
    ctx.update(&[key('g'), key('l'), key('l')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn log_other() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "this-should-be-at-the-top", "");
    commit(ctx.dir.path(), "this-should-not-be-visible", "");

    ctx.update(&[key('g'), key('l'), key('l'), key('j'), key('l'), key('o')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn show() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "firstfile", "This should be visible\n");
    ctx.update(&[key('g'), key('l'), key('l'), key_code(KeyCode::Enter)]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn rebase_conflict() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "new-file", "hello");

    run(ctx.dir.path(), &["git", "checkout", "-b", "other-branch"]);
    commit(ctx.dir.path(), "new-file", "hey");

    run(ctx.dir.path(), &["git", "checkout", "main"]);
    commit(ctx.dir.path(), "new-file", "hi");

    run(ctx.dir.path(), &["git", "checkout", "other-branch"]);
    run(ctx.dir.path(), &["git", "rebase", "main"]);

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn merge_conflict() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "new-file", "hello");

    run(ctx.dir.path(), &["git", "checkout", "-b", "other-branch"]);
    commit(ctx.dir.path(), "new-file", "hey");

    run(ctx.dir.path(), &["git", "checkout", "main"]);
    commit(ctx.dir.path(), "new-file", "hi");

    run(ctx.dir.path(), &["git", "merge", "other-branch"]);

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn moved_file() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "new-file", "hello");
    run(ctx.dir.path(), &["git", "mv", "new-file", "moved-file"]);

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn hide_untracked() {
    let mut ctx = TestContext::setup_clone(60, 10);
    let mut config = ctx.state.repo.config().unwrap();
    config.set_str("status.showUntrackedFiles", "off").unwrap();
    run(ctx.dir.path(), &["touch", "i-am-untracked"]);

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn new_commit() {
    let mut ctx = TestContext::setup_clone(60, 10);
    commit(ctx.dir.path(), "new-file", "");

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn push() {
    let mut ctx = TestContext::setup_clone(60, 10);
    commit(ctx.dir.path(), "new-file", "");

    ctx.update(&[key('P'), key('p')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn fetch_all() {
    let mut ctx = TestContext::setup_clone(60, 10);
    clone_and_commit(&ctx.remote_dir, "remote-file", "hello");

    ctx.update(&[key('f'), key('a')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn pull() {
    let mut ctx = TestContext::setup_clone(60, 10);
    clone_and_commit(&ctx.remote_dir, "remote-file", "hello");

    ctx.update(&[key('F'), key('p')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}
