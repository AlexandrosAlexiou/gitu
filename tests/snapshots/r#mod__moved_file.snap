---
source: tests/mod.rs
expression: ctx.redact_buffer()
---
Buffer {
    area: Rect { x: 0, y: 0, width: 60, height: 20 },
    content: [
        "ðŸ¢’On branch main                                             ",
        "â–ŒYour branch is ahead of 'origin/main' by 1 commit.         ",
        "                                                            ",
        " Staged changes (2)                                         ",
        " moved-file                                                 ",
        " @@ -0,0 +1 @@                                              ",
        " +hello                                                     ",
        " new-file                                                   ",
        " @@ -1 +0,0 @@                                              ",
        " -hello                                                     ",
        "                                                            ",
        " Recent commits                                             ",
        " _______ main add new-file                                  ",
        " _______ origin/main add initial-file                       ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: BOLD,
        x: 1, y: 0, fg: Yellow, bg: Reset, underline: Reset, modifier: BOLD,
        x: 15, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: BOLD,
        x: 0, y: 1, fg: Blue, bg: Reset, underline: Reset, modifier: DIM,
        x: 1, y: 1, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 3, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 15, y: 3, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 4, fg: Magenta, bg: Reset, underline: Reset, modifier: NONE,
        x: 11, y: 4, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 5, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 14, y: 5, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 6, fg: Green, bg: Reset, underline: Reset, modifier: NONE,
        x: 7, y: 6, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 7, fg: Magenta, bg: Reset, underline: Reset, modifier: NONE,
        x: 9, y: 7, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 8, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 14, y: 8, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 9, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 7, y: 9, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 11, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 15, y: 11, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 12, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 8, y: 12, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 9, y: 12, fg: Green, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 12, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 13, fg: Yellow, bg: Reset, underline: Reset, modifier: NONE,
        x: 8, y: 13, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 9, y: 13, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 13, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
